# CCB Python Bootcamp ("PyCamp") üêç
## A biannual introductory workshop to using Python for bioinformatics

Presented by the [Center for Computational Biology](ccb.berkeley.edu). Please contact `ccbadmin(at)berkeley.edu` for all questions.

--------------

## What is this?
This is a repository containing the notebooks, datasets, and visual assets for CCB's Python Bootcamp.

## Table of contents
- [Lecture notebooks](#notebook-contents)
- [Zoom](#zoom)
- [Datasets](#datasets)
- [FAQs](#FAQs)
- [Authorship](#authorship)
- [License](#license)

## Zoom
Link to the zoom is [here](https://berkeley.zoom.us/s/97071151872). Lectures are recorded and will be posted on the PyCamp slack when they are ready!

## Slack

Make sure to sign up for our slack [here](https://join.slack.com/t/slack-qxv2392/shared_invite/zt-2o1o6960x-2vTSo5CW5ezl_fTS7ymEaQ). Lecturers and TAs will be mainly communicating through our slack.

## Lecture notebooks
Each notebook is intended to be a lesson on a particular unit of Python. 

| Day | Content | Lecture Notebooks | Cheat sheet |
| ---- | ------ | ------ | ------ |
| **Day 1** (Monday) | Interacting with Google Colaboratory. Introduction to types, lists, tuples, sets, dictionaries. Indexing and slicing iterables. Built-in functions. Boolean logic and simple control flow. | [Monday lecture](https://drive.google.com/drive/folders/1M6ArM2uBbPkgdBLFcvr3K2EpKzS5zVcq?usp=sharing) |
| **Day 2** (Tuesday) | Review basic data structures and methods. Writing custom functions. Introduction to external packages (`numpy`). | [Tuesday lecture](https://drive.google.com/drive/folders/1JU_ItTNads_KM3rzfO_Zay2hOf5KD7OV?usp=sharing) |
| **Day 3** (Wednesday) | Review and cover more detail on arrays. Mini-projects on data cleaning, exploration, and basic visualization with the `badhealth` and `hepatocellular` datasets. | [Wednesday lecture](https://drive.google.com/drive/folders/174SG_XVhjFMHnRfYza7ozw3zPjucbM61?usp=sharing) |
| **Day 4** (Thursday) | Introduction to `pandas` operations: indexing, slicing, querying, filtering, merging. Exploration of the `clinvar` dataset.| [Thursday lecture](https://drive.google.com/drive/folders/1LbBEDQVrZc2KWXXZxOzPiixyMTpUUGly?usp=sharing) |
| **Day 5** (Friday) | Final mini-project: synthesizing content to analyze a subset of the [Tabula Muris Senis](https://www.nature.com/articles/s41586-020-2496-1) dataset.| [Friday lecture](https://drive.google.com/drive/folders/1aVRhVEBH-hdPSKcC41BFb9lYfBj4g-6A?usp=sharing) |

## Datasets
All datasets used in these materials are derived from publicly available data.

| Dataset | Source |
| ---- | ------ |
| `badhealth` | Originally from `Rdatasets`, downloaded [here](https://vincentarelbundock.github.io/Rdatasets/).
| `hepatocellular` | Originally from `Rdatasets`, downloaded [here](https://vincentarelbundock.github.io/Rdatasets/).
| `clinvar` | A table of variants from the [NCBI ClinVar](https://www.ncbi.nlm.nih.gov/clinvar/) database. Curated by [Andrew Sharo](https://www.andrewsharo.com/). |
| `facs`, `metadata_facs`, `metadata_droplet` | Single-cell count matrix and metadata tables from the [Tabula Muris Senis](https://www.nature.com/articles/s41586-020-2496-1) dataset. Downloaded using the [UCSC Cell Browser](https://cells.ucsc.edu/?ds=tabula-muris-senis). |

## FAQs
Find answers [here](https://ccb.berkeley.edu/ccb-bioinformatics-bootcamp-january-2022-faq/).

## Authorship
Current notebooks and curriculum were developed by [Stacy Li](stacy.li) in Summer 2022, with exercises contributed by PyCamp staff in the years following. Pre-2022 notebooks were developed by previous generations of PyCamp staff.

## License
This repository is licensed under the [CC 4.0 license](https://creativecommons.org/licenses/by/4.0/). You are free to share, redistribute, and adapt these notebooks, although we request that you provide a link to this repository.
