{"cells":[{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":15467,"status":"ok","timestamp":1704406024864,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"OVegqce_rzV7","outputId":"0e83669e-9c17-43bb-e31e-65d3e8390ff2"},"outputs":[],"source":["# make sure to run this cell to import the external files we need for today\n","# and load in the appropriate packages\n","\n","!git clone https://github.com/ccbskillssem/pythonbootcamp.git\n","\n","import numpy as np\n","import pandas as pd\n","import matplotlib.pyplot as plt"]},{"cell_type":"markdown","metadata":{"id":"ESngQKLsgTnt"},"source":["# Introduction\n","\n","Welcome to our final day of PyCamp! Today, you'll be embarking on your final mini-project, which will unify concepts and themes you've learned over the past several days."]},{"cell_type":"markdown","metadata":{"id":"mbHc28GXmyPc"},"source":["## Single-cell in a nutshell\n","\n","Recall the last biology class that you took: you probably learned at some point the **central dogma of molecular biology**, which is a fancy name for the principle that DNA gives rise to RNA, and RNA gives rise to protein.\n","\n","Scientists are able to survey DNA, RNA, and proteins using various methods: **sequencing**, the process of recording the nucleic acid sequence of a DNA/RNA molecule, is quite popular due to its relative scalability and cost efficiency.\n","* DNA sequencing gives us information on the organism's **genome**, akin to the blueprint of its biology. (We will not be discussing this today, sorry!)\n","* **RNA sequencing** gives us insight into the creation of an organism's biology: what elements of the blueprint are being realized?\n","\n","**Single-cell RNA sequencing** (scRNA-seq) is an increasingly popular method for capturing and assigning identity (gene origin) to RNA molecules circulating within a single cell. Briefly, tissue samples are dissociated into single cells, and each RNA molecule within a cell is labeled with a barcode of sorts that indicates its cellular origin. The number and identity of RNA molecules per cell is then determined computationally. This method yields high(er) resolution insights into tissue heterogeneity and temporal cellular processes, such as cell differentiation.\n","\n","> *Note*: On Monday, we said that our goal isn't to teach you specific bioinformatic analyses, and that's still true. **Today's mini-project is not a full pipeline for analyzing single-cell data**: however, you will be much better equipped to learn how to do that once you're done with the mini-project. (Finish the whole mini-project and you'll realize how much is in your grasp right now!)\n","\n"]},{"cell_type":"markdown","metadata":{"id":"UAfpenHpliVD"},"source":["## What is a cell atlas?\n","\n"," A **cell atlas** is an analyzed single-cell dataset comprising multiple cell types over some type of biological scale. Cell atlases grant insight into  cellular heterogeneity across anatomical and physiological scales, and they are especially useful for studying healthy development (ex. stem cell differentiation) and disease progression (ex. cancer).\n","\n"," The [Tabula Muris Senis](https://www.nature.com/articles/s41586-020-2496-1) is a cell atlas of aging mouse tissues, comprised of samples from multiple healthy mouse specimens of varying ages. For our mini-project, we'll use a subset of the cell atlas data and attempt to replicate some of the findings from the linked paper."]},{"cell_type":"markdown","metadata":{"id":"-tNQwydVljw7"},"source":["## Overview\n","\n","We want to work the Tabula Muris Senis data because it has both real biological meaning (spoiler alert: gene expression does change with aging!) *and* is natively represented in a tabular format, much like the arrays and DataFrames you've become accustomed to handling in the last couple of days.\n","\n","We're going to examine the bladder dataset, as the authors found that bladder tissues change quite significantly with age.\n","\n","> \"*The bladder shows pronounced changes in cell-type composition with age: whereas the mesenchymal compartment of this tissue decreases by a factor of three over the lifetime of the mouse, the urothelial compartment increases by a similar amount...*\"\n","\n","This dataset contains two tables:\n","\n","1. **A metadata table**: This will be a DataFrame that describes auxilary features about our data: for example, what was the age of each mouse? What tissue did our sample come from, and how many RNA molecules were captured in each sample? This information can be quite useful for guiding what we do with the data.\n","2. **A count matrix**: This will be a DataFrame that describes *gene expression* in each of our cells. This means that we'll have information on how many \"instances\" of a gene (read: RNA molecules) were found per cell. We should expect that some genes will be highly expressed (many RNA molecules) in some cells, but near absent in others.\n","\n","Let's go ahead and get started!"]},{"cell_type":"markdown","metadata":{"id":"xDPNOQfKgVXm"},"source":["## Downloading the data\n","\n","The Tabula Muris Senis data is hosted on an external server: we can download a local copy of our desired data using a command line tool called `wget`. Although we don't have access to a terminal in the free version of Colab, we can still execute command line commands using **magic commands**.\n","\n","1. **For a single line, use the `!` line magic command at the beginning of the line**: This tells the notebook that the following line should be run via the comand line.\n","2. **For multiple lines, use the `%%bash` magic command at the beginning of the cell.** This tells the notebook that all lines following `%%bash` need to be run via the command line.\n","\n","These commands work in both Colab notebooks and Jupyter notebooks, and they're very convenient for documenting your full analysis workflow. You can read about other useful notebook magic commands [here](https://ipython.readthedocs.io/en/stable/interactive/magics.html)."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":2704,"status":"ok","timestamp":1704406027565,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"WNaxPd-TlLtT","outputId":"75f595f7-4042-407d-e325-b2bf639efeb8"},"outputs":[],"source":["# run this cell to download: this should only take a few seconds\n","%%bash\n","wget https://cells.ucsc.edu/tabula-muris-senis/facs/bladder/exprMatrix.tsv.gz -O bladder-facs.tsv.gz --no-check-certificate --no-verbose\n","wget https://cells.ucsc.edu/tabula-muris-senis/facs/bladder/meta.tsv -O bladder-facs.meta.tsv --no-check-certificate --no-verbose"]},{"cell_type":"markdown","metadata":{"id":"fPRbeR7LaML3"},"source":["# Review & warm-up\n","\n","We'll begin with a series of review exercises that we'll do together. Afterwards, you'll have time to work on the mini-project."]},{"cell_type":"markdown","metadata":{"id":"e4lOOSQl2DRM"},"source":["## Examining the metadata\n","Yesterday, you learned how you can perform many familiar Excel-like functions using `pandas` DataFrame functions and methods. We'll have you briefly refresh these skills by examining our metadata.\n","\n","Our metadata is stored in a simple `.tsv` file called `bladder-facs.meta.tsv`, which we'll load in below:"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":417},"executionInfo":{"elapsed":170,"status":"ok","timestamp":1704406027733,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"Y7emflW9pGe9","outputId":"715857b5-6bbc-4c5c-c911-be78022ff661"},"outputs":[],"source":["# try it out:\n","# use the appropriate pandas function to load in the metadata table and assign it to a variable called metadata_facs\n","# then use the .head() method to preview the first few rows of the table\n"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":4,"status":"ok","timestamp":1704406027734,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"Ua4gRMF9pJ2E","outputId":"986f9ff7-424a-4ac8-ab6f-2dbec0117f2a"},"outputs":[],"source":["# try it out:\n","# inspect the number of rows and columns in this table\n"]},{"cell_type":"markdown","metadata":{"id":"Qe_iH0jhpLhO"},"source":["Each row of `metadata_facs` contains the metadata of a single cell in the dataset. The `index` column contains a unique string \"name\" for each cell: let's use this as the actual row Index (column of row labels) for the DataFrame."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":448},"executionInfo":{"elapsed":187,"status":"ok","timestamp":1704406027918,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"nzmU6I_xpNA2","outputId":"5960f036-8683-4aeb-f8bc-7c2083015850"},"outputs":[],"source":["# try it out:\n","# use the appropriate method to set the index of metadata_facs to the index column\n","# then use .head() to check that it worked\n"]},{"cell_type":"markdown","metadata":{"id":"IAXrwJt2pSU9"},"source":["Now, let's review a couple of very useful methods that you'll need for today's mini-project.\n","\n","Although we know that each row represents one cell, we still don't know how many samples are represented in this dataset, nor do we know the identity and composition (cell number, cell type) of each sample.\n","\n","`.unique()` and `.value_counts()` are powerful methods that summarize the unique values of a categorical column and the number of value counts associated with those unique values.\n","\n","* `.unique()` is a method for **Series**: it does not require any input, and it returns an array of unique values contained in the Series. This method works well when you don't need counts, you only need values.\n","* `.value_counts()` is a method for **DataFrames**: it takes a variable name and returns a Series with aggregated value counts for each unique value. By default, the Series will be sorted in ascending order of counts.\n","  * **Note**: By default, this method excludes `NA` values from the values to be counted. You can include `NA` values using the `dropna = False` parameter."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":145,"status":"ok","timestamp":1704406028061,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"FqXRyAUxpVdS","outputId":"2b148bb6-479e-4aef-c4e2-2bf318d3086e"},"outputs":[],"source":["# try it out:\n","# access the column names of metadata_facs\n"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":3,"status":"ok","timestamp":1704406028061,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"Losg7V0PpYbx","outputId":"150e78f0-f148-477f-c276-2311c0bc95d1"},"outputs":[],"source":["# try it out:\n","# use .unique() to get unique values in mouse.id\n"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":2,"status":"ok","timestamp":1704406028061,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"H9caPECKpbH8","outputId":"b985e91e-5de7-44fc-bbab-f9a2b8b765f7"},"outputs":[],"source":["# try it out:\n","# use .value_counts() to get the number of cells per mouse.id\n"]},{"cell_type":"markdown","metadata":{"id":"0icen28TpepA"},"source":["`.value_counts()` is especially useful if you want to aggregate counts across multiple variables.\n","\n","There are two columns in `metadata_facs` that will can tell us about cellular identity: `cell_ontology_class` and `free_annotation`. `cell_ontology_class` tells us about the anatomical placement of the cell, and `free_annotation` tells us about the specific cell type.\n","\n","We'll revisit this later in the project, but for now, let's use this as an opportunity to test out `.value_counts()` for multiple variables."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":3,"status":"ok","timestamp":1704406028202,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"RuKSQ6MipeQT","outputId":"147c0ffd-be23-4418-c521-f28fd8da61f7"},"outputs":[],"source":["# try it out:\n","# find the number of cells in each sample for each (anatomical) cell type\n","\n","# bonus:\n","# this Series would be more readable if each sample's counts were shown ordered\n","# by mouse.id, rather than ordered by count values. how might you do this?\n"]},{"cell_type":"markdown","metadata":{"id":"nxYToutQpqja"},"source":["## Plotting the metadata\n","\n","For the past couple of days, you've been using `pyplot` to create simple plots from arrays and DataFrame columns.\n","\n","Let's start off with a quick review: let's say that we want to create a bar plot with the `sex` categorical on the x axis and cell counts for each value of `sex` on the y axis.\n","\n","We can use a function called `plt.bar()` to create a simple bar plot. `plt.bar()` requires two inputs:\n","1. `x`: A list, array, or Series with categories to be plotted on the x axis. These will be the labels for each bar.\n","2. `height`: A list, array, or Series with numerical values to be plotted on the y axis.\n","\n","Different plotting functions will require different inputs, depending on the type of plot: for example, `plt.scatter()` wouldn't take `height` as an input, it would require `y` as an input.\n","> It's almost essential to keep the documentation for your plotting package open in a separate tab. Remember, *don't memorize*, just know where to look for answers."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":448},"executionInfo":{"elapsed":778,"status":"ok","timestamp":1704406028978,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"M87cDaOCppwM","outputId":"2c8b9393-ab70-47f8-89f5-30488bc1f5f8"},"outputs":[],"source":["# try it out:\n","# use plt.bar() to create the aforementioned plot\n","# check the y height of each bar against the categorical counts in the previous cell\n"]},{"cell_type":"markdown","metadata":{"id":"t-0FA68ZjN9R"},"source":["We can adjust the size of our plots using `plt.figure()` with the `figsize` parameter. `figsize` accepts a tuple of integers in the format of `(width, height)`."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":386},"executionInfo":{"elapsed":670,"status":"ok","timestamp":1704406029646,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"C1MjAP6yjY3V","outputId":"ea481097-56c5-42cf-e021-d1d4f8d5e56b"},"outputs":[],"source":["# try it out:\n","# copy your plt.bar() command from the previous cell\n","# set plt.figure() with a figsize tuple of your choice\n"]},{"cell_type":"markdown","metadata":{"id":"B7qRr9MCpwnv"},"source":["So far, we've only made plots to help us inspect our data. If we want to make presentation-worthy plots, we need to add labels.\n","\n","* `plt.title()`: Takes a string as input and labels the top of the plot with the string. Takes an optional `size` parameter (ex. `size = 16`) to adjust the title size.\n","* `plt.xlabel()` and `plt.ylabel()`: Same as above, but labels the x and y axes respectively."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":430},"executionInfo":{"elapsed":1078,"status":"ok","timestamp":1704406030723,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"iWnOj0ahpyWe","outputId":"01cc0fad-8f8b-43a4-c9e1-54ec1fbdef4d"},"outputs":[],"source":["# try it out:\n","# copy your code from the previous cell\n","# use the plt functions shown above to add plot labels\n"]},{"cell_type":"markdown","metadata":{"id":"YX0zFqWDqLOi"},"source":["It's fairly rare that we only want to plot the counts of one categorical. For example, what if we wanted to create a grouped bar chart showing cell counts split by sample sex and age?\n","\n","It's time for us to introduce `seaborn`, a package that extends the plotting functionalities of `pyplot`."]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":3447,"status":"ok","timestamp":1704406034168,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"4rFthumPqcg0"},"outputs":[],"source":["# make sure to run this cell!\n","import seaborn as sns"]},{"cell_type":"markdown","metadata":{"id":"-dDcDasexOdo"},"source":["`seaborn` provides useful functions that can simultaneously aggregate and plot data directly from DataFrames. It's exceptionally powerful for visualizing mixed categorical and numerical data, which is what we want to do today.\n","<br><br>\n","\n","<img src = 'https://seaborn.pydata.org/_images/function_overview_8_0.png' align = 'center'>\n","<br><br>\n","\n","`seaborn` functions are built with DataFrames in mind. Most functions tend to take similar key inputs:\n","* `data`: The DataFrame containing your data of interest. **This is always the first input to a `seaborn` function.**\n","* `x`: A variable to be plotted on the x-axis.\n","* `y`: A variable to be plotted on the y-axis.\n","* `hue`: A variable for coloring groups in the plot."]},{"cell_type":"markdown","metadata":{"id":"t60bYU900YWu"},"source":["Let's return to our earlier prompt: we want to create a grouped bar chart showing cell counts split by sample sex and age. Let's start by calculating the counts we expect:"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":3,"status":"ok","timestamp":1704406034168,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"u0zVQHnz2F6n","outputId":"a4a37b0b-15c9-4eb0-b2f1-1e1c81564470"},"outputs":[],"source":["# try it out:\n","# use the same .value_counts() method to get cell counts for each combination of sex and age\n"]},{"cell_type":"markdown","metadata":{"id":"71d-RdQD3Ojq"},"source":["\n","`sns.countplot()` takes a column of values and calculates counts, plotting them as bars. It uses the `hue` parameter to subdivide bars, and we can use it to create our desired subdivided bar chart."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":492},"executionInfo":{"elapsed":1360,"status":"ok","timestamp":1704406035526,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"_51WqgpEqiCS","outputId":"6eacb202-5a78-4373-b557-0a453022ecd7"},"outputs":[],"source":["# try it out:\n","# use sns.countplot() with two categorical variables: sex and age\n","# we want to plot sex on the x axis and subdivide bars by age\n","# make sure to add labels and a title\n"]},{"cell_type":"markdown","metadata":{"id":"jboYMGrvnfZ5"},"source":["It seems like cells from 3 month old male samples are overrepresented in this dataset. One reason might be that some samples yielded more cells than others.\n","\n","Using `sns.countplot()`, create a plot that:\n","- Shows total cell count per `mouse.id`.\n","- Uses the optional `palette` parameter with a palette of your choosing: you can view different options [here](https://www.practicalpythonfordatascience.com/ap_seaborn_palette). (We recommend the classic `viridis` or Berkeley blue-and-gold `cividis`!)\n","\n","> *Bonus*: Use the optional `order` parameter to specify the order of `mouse.id` values: the specimen with the greatest number of total cells should be first, and the rest of the specimens should follow in descending order."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":467},"executionInfo":{"elapsed":1801,"status":"ok","timestamp":1704406037326,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"cIdpe_txnQZD","outputId":"dcd0e2be-b06d-41b6-85c9-5ee54d947a2c"},"outputs":[],"source":["# try it out:\n","# create the plot described above\n","# note: sns.countplot() takes either x or y as input\n","# x creates a horizontal plot, y creates a vertical plot\n"]},{"cell_type":"markdown","metadata":{"id":"xEcsWx7X5O5i"},"source":["# Mini-project: Tabula Muris Senis bladder data\n","\n","It's time to start on the mini-project!\n","\n","This project will require you to revisit topics you've learned in the past few days, emphasizing Wednesday and Thursday's content. We'll also introduce concepts and packages you haven't seen before: consider this a bridge from PyCamp to the rest of your work in Python!\n","\n","**We highly encourage you to work in groups.** TAs will be available to help you as you go, and we'll pause for breaks and lecturer-led Q&A as we go along.\n","\n","Due to the nature of this project, **this is the only notebook for this session.** If your Colab notebook disconnects at any point, you can restore your progress by using the `Runtime -> Run before` option in the top (horizontal) menu."]},{"cell_type":"markdown","metadata":{"id":"3RHCrTTY2NC1"},"source":["## Examining gene expression\n","The second table in our dataset is a **count matrix**, with gene names on one axis and cell names on the other. Gene expression is denoted by **counts**: each cell will have a different number of counts for each gene, scaling with the number of RNA molecules identified in the cell.\n","\n","Below, use the appropriate `pandas` function to load in the count matrix (`bladder-facs.tsv.gz`) and assign it to a variable called `facs`, then inspect it with `.head()`.\n","\n","> Note: `facs` is a large DataFrame, and it may take a few seconds to load."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":273},"executionInfo":{"elapsed":14892,"status":"ok","timestamp":1704406052216,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"RMThmih8EV8w","outputId":"7af590de-d7c0-4740-8384-07037e383539"},"outputs":[],"source":["### write your code below ###\n"]},{"cell_type":"markdown","metadata":{"id":"KKxU5L22eD6O"},"source":["The first column of `facs` is a column with the label `gene`, containing gene names. The remaining columns are labeled with cell names, containing counts for each `gene`.\n","\n","We want this count matrix layout to match our metadata table layout, where each cell is represented by one labeled row: thus, we'll need to **transpose** `facs`. We originally learned about transposition for arrays, and it works largely the same for DataFrames: **you will need to look up the appropriate method for this in the `pandas` documentation.**\n","\n","Here are the steps you need to take after you have the correct function in hand:\n","1. Set the `gene` column as the row Index: after transposition, `gene` will be used for column labels.\n","2. Transpose (switch) the axes of the table using the appropriate DataFrame method.\n","\n","Make sure that you're updating `facs` with each of these changes. If you've done everything correctly, `facs` should have the same number of rows as `metadata_facs`."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":342},"executionInfo":{"elapsed":5,"status":"ok","timestamp":1704406052216,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"VDooGzdnK4T-","outputId":"a9129dc5-beef-436d-a5cc-f7ecb8f5ce9a"},"outputs":[],"source":["### write your code below ###\n"]},{"cell_type":"markdown","metadata":{"id":"B1PzJPz1w6BX"},"source":["This dataset has already undergone standard quality control and normalization steps for scRNA-seq data: that means that we don't need to do the same data cleaning steps that we did with the other mini-projects. Instead, we'll get right to examining gene expression values across different genes."]},{"cell_type":"markdown","metadata":{"id":"LSaX8Uqat156"},"source":["### Single distributions\n","\n","Let's start out with a fairly simple task: visualizing the distribution of a single gene.\n","\n","Single distributions are commonly visualized using histograms, which we've created before using `plt.hist()`. Below, try using the very similar `sns.histplot()` to create a histogram of gene expresssion values for `Fmo5`."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":490},"executionInfo":{"elapsed":645,"status":"ok","timestamp":1704406052858,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"Wsl2mVEA_MqO","outputId":"ba7b4074-a072-4822-9e42-57e780c96b8f"},"outputs":[],"source":["### write your code below ###\n"]},{"cell_type":"markdown","metadata":{"id":"7MRoNoGVwB5R"},"source":["We can also use boxplots to visualize gene expression distributions: copy the code you used above, but replace `sns.histplot()` with `sns.boxplot()`."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":490},"executionInfo":{"elapsed":200,"status":"ok","timestamp":1704406053056,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"BqNWrSBTwOqs","outputId":"2b0195ae-4dee-48b3-fe58-fde0a98d0adc"},"outputs":[],"source":["### write your code below ###\n"]},{"cell_type":"markdown","metadata":{"id":"IVaGnZSEteeq"},"source":["### Multiple distributions\n","There are several different ways to visualize multiple distributions at the same time. We'll use the plot types in the [Fundamentals of Data Visualization](https://clauswilke.com/dataviz/boxplots-violins.html), which are all available in `seaborn`.\n","\n","As we mentioned before, `seaborn` is built for DataFrames and can make basic assumptions about how to plot your data. This is particularly useful for plotting multiple distributions.\n","\n","Given a DataFrame of gene expression values, `seaborn` will automatically plot both columns and color them by their column labels.\n","* We've provided a list of genes called `subset_genes` below: select these columns from `facs` and assign the resultant DataFrame to a new variable called `subset_facs`.\n","* Provide `subset_facs` as the sole input to `sns.histplot()`.\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":448},"executionInfo":{"elapsed":814,"status":"ok","timestamp":1704406053867,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"KPfSRwMEtV7s","outputId":"5abbc866-1eca-4fcb-f7b1-cfa464ff8cb9"},"outputs":[],"source":["subset_genes = ['Fmo5', 'Id3', 'Lgals3', 'Wfdc2', 'Pcolce']\n","\n","### write your code below ###\n"]},{"cell_type":"markdown","metadata":{"id":"77GdgGiEynA8"},"source":["As you can see, histograms can become a bit visually busy with multiple distributions. Boxplots & boxplot variants (violin plots, boxen plots, strip plots, and swarm plots) tend to work much better for higher numbers of distributions visualized in the same plot.\n","\n","Take a moment to see what these plots look like: simply substitute `sns.histplot()` with the functions below.\n","* `sns.boxplot()`\n","* `sns.violinplot()`\n","* `sns.boxenplot()`\n","\n","The plots below work best for smaller datasets: we recommend using `subset_facs.sample(50)` to avoid warnings about overloading the plot.\n","* `sns.stripplot()`: This plot works best for mid-sized datasets: we recommend using `subset_facs.sample(500)` to avoid overloading the plot.\n","* `sns.swarmplot()`: This plot works best for smaller datasets: we recommend using `subset_facs.sample(25)` to avoid overloading the plot."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":524},"executionInfo":{"elapsed":529,"status":"ok","timestamp":1704406054392,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"4L8SAgixy_Yv","outputId":"47d14a7f-531d-491c-c8bb-398edfa74d2f"},"outputs":[],"source":["### write your code below ###\n"]},{"cell_type":"markdown","metadata":{"id":"1WyeEUkH___L"},"source":["### Subdivided distributions\n","\n","Not all genes are expressed uniformly across all cells, especially cells of different types. What if we wanted to examine expression of these genes in different cell types? In order to do this, we'll need to *merge* our two DataFrames together into a super-DataFrame that contains all the information we want to use.\n","\n","`facs` and `metadata_facs` can be merged using the row Index (row labels) of each DataFrame, which contains shared values (cell names). However, there's one small problem: the `facs` row Index doesn't actually have a name."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":137,"status":"ok","timestamp":1704406054527,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"PbetHXNEvb5c","outputId":"30b9b228-02dc-4bcb-f519-d8232639ce42"},"outputs":[],"source":["print('metadata_facs:', metadata_facs.index.name)\n","print('facs:', facs.index.name)"]},{"cell_type":"markdown","metadata":{"id":"09pHUb0Cz-IQ"},"source":["Recall that we can merge by providing a list of DataFrames/Series to `pd.merge()`, or by calling the `.merge()` method on an existing DataFrame. In either case, we need to specify the **merge key**: a column or index that has the same name in each DataFrame.\n","\n","Below, perform the following operations:\n","* Update the `facs` Index name to match the `metadata_facs` index name.\n","* Merge `subset_facs` with `metadata_facs` using `'index'` as the merge key.\n","  * Save the resultant DataFrame to a new variable called `merged_subset`."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":478},"executionInfo":{"elapsed":4,"status":"ok","timestamp":1704406054527,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"Dg0_YqrL0KVk","outputId":"41b659ac-a37d-4fa4-f279-ee3c6e453c3b"},"outputs":[],"source":["### write your code below ###\n"]},{"cell_type":"markdown","metadata":{"id":"ArMO89lI0dwk"},"source":["Using `merged_subset`, create a boxplot of gene expression for `Id3`, then subdivide the boxplot by `age` and `free_annotation`."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":467},"executionInfo":{"elapsed":254,"status":"ok","timestamp":1704406054778,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"PX7Hy6ndshE5","outputId":"4d84434b-6c0e-45ca-80b4-37851ee94d3f"},"outputs":[],"source":["### write your code below ###\n"]},{"cell_type":"markdown","metadata":{"id":"AWAOAREcigil"},"source":["## Principal component analysis"]},{"cell_type":"markdown","metadata":{"id":"PTBzUFDNUNj1"},"source":["**Principal component analysis** is a technique commonly used in scRNA-seq data pre-processing, exploration, and visualization.\n","\n","> *Principal component analysis (PCA) ... is accomplished by linearly transforming the data onto a new coordinate system (principal components) such that the directions capturing the largest variation in the data can be easily identified. Many studies use the first two principal components in order to plot the data in two dimensions and to visually identify clusters of closely related data points.*  - from [Wikipedia](https://en.wikipedia.org/wiki/Principal_component_analysis)"]},{"cell_type":"markdown","metadata":{"id":"h5_vhETDlQLp"},"source":["### Feature selection\n","\n","PCAs are typically used to summarize variation and reduce overall dimensionality in the dataset. scRNA-seq count matrices frequently have many thousands of dimensions (genes), with some genes contributing more to variation than others: these are, as you might predict, **highly variable genes** (HVGs). Filtering the count matrix to only contain highly variable genes is a process called **feature selection**"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":4,"status":"ok","timestamp":1704406054778,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"JgIzM4wilRdB","outputId":"7b1c8f31-380f-44cd-e6d6-5e65bdd17523"},"outputs":[],"source":["# identification of highly variable genes is beyond the scope of PyCamp\n","# so we've provided them for you here (an array of HVG gene names)\n","hvg = pd.read_csv('pythonbootcamp/day_5/hvg.csv')['genes'].values\n","hvg"]},{"cell_type":"markdown","metadata":{"id":"zg06Yeq8PNv1"},"source":["From `facs`, select the columns in `hvg`: assign the resultant DataFrame to a new variable called `facs_hvg`."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":374},"executionInfo":{"elapsed":166,"status":"ok","timestamp":1704406054942,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"PieRDVcmlUSX","outputId":"a78afb55-6895-4cc2-c793-976d7f5ca1ec"},"outputs":[],"source":["### write your code below ###\n"]},{"cell_type":"markdown","metadata":{"id":"7ORAbS5VODzy"},"source":["### Dimensionality reduction\n","\n","We need to bring in another package to perform our PCA: `sklearn` is short for **scikit-learn**, a popular Python package used for machine learning applications.\n","\n","We only need a single function from `sklearn`:\n","* `sklearn.decomposition.PCA()`: Creates a PCA generator object akin to the random number generator object created by `np.random.default_rng()`. Takes two inputs: the `n_components` parameter specifies the number of **principal components** (PCs) to be returned, and the `random_state` parameter allows you to set a seed value for the pseudo-random aspect of PCA.\n","\n","Run the cell below to create the PCA generator object: we've set the random state to ensure that your results are comparable to the answer key."]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":387,"status":"ok","timestamp":1704406055328,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"Ed-grubdTfpc"},"outputs":[],"source":["# make sure to run this cell! nothing will work without it\n","\n","import sklearn.decomposition\n","pca = sklearn.decomposition.PCA(n_components = 50, random_state = 7)"]},{"cell_type":"markdown","metadata":{"id":"iIL97gxyX2M-"},"source":["`pca` is a generator that will transform a dataset and summarize it into an array of the first 50 principal components. The first principal component (PC1) will summarize the largest amount of variation in the data, the second (PC2) will summarize the second largest amount of variation, and so on and so forth.\n","\n","We need to use our new `pca` object to fit and transform `facs_hvg`: we can accomplish this by chaining together two methods:\n","* `.fit()`: Takes a DataFrame or array as input, fits `pca` to the input data.\n","* `.transform()`: Takes a DataFrame or array as input, transforms the input data according to `pca` into an array. **This method can only be used with the same input provided to `.fit()`**."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":1004,"status":"ok","timestamp":1704406056330,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"o_LfcTzIXioE","outputId":"98a715f1-7f22-40db-aa9d-59d31cfbdfcf"},"outputs":[],"source":["### write your code below ###\n"]},{"cell_type":"markdown","metadata":{"id":"-9FPqAA8XBV6"},"source":["This array contains the transformed principal component coordinates of each row in `facs`. However, this array is currently unlabeled, which will make it difficult for us to use in plotting: we need to transform it into a DataFrame with an appropriate index and column names.\n","\n","Construct a DataFrame from `pca_array` using the `pd.DataFrame()` function, then assign it to a variable called `pca_df`. You will need to provide column labels and an index using the appropriate parameters, which you can find in the [documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html).\n","\n","*Hint*: The PCA transformation reduces the number of columns, but the number and order of rows, as well as the row labels, remains unchanged."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":562},"executionInfo":{"elapsed":4,"status":"ok","timestamp":1704406056330,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"JaAmabXbb-GF","outputId":"5e8f10a5-c03b-4a5b-fd07-554d8a0801c2"},"outputs":[],"source":["### write your code below ###\n"]},{"cell_type":"markdown","metadata":{"id":"k2RsPTzEaI3R"},"source":["## PCA visualization\n","As Wikipedia mentioned, it's quite common to use the first two principal components for exploratory visualization, as they capture the maximum amount of variation in the dataset in two dimensions.\n","\n","Start off by creating a basic scatterplot of the first two PCs with `sns.scatterplot()`.\n","> *Note*: PCA plots typically hide coordinate values and ticks: you can remove x and y ticks on PCA plots by adding `plt.xticks([], [])` and `plt.yticks([], [])`. All plots in the answer key will obey this convention."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":444},"executionInfo":{"elapsed":499,"status":"ok","timestamp":1704406056826,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"H1gDT7gmexqL","outputId":"05591938-79fe-4cba-d13f-f94b6d864572"},"outputs":[],"source":["### write your code below ###\n"]},{"cell_type":"markdown","metadata":{"id":"FBorHYVTpixL"},"source":["### Metadata\n","PCA plots are typically colored by some metadata variable that may be relevant for understanding why data points cluster together: what if we simply added principal components as extra columns on our metadata table?\n","\n","We can take advantage of the fact that `metadata_facs`, `facs`, and `facs_df` all use the same index: using the `.merge()` method, we can create a single convenient DataFrame containing the first two PCs and the metadata for each cell.\n","\n","Select the first two PCs in `pca_df` and join them with `metadata_facs` using `.merge()`, assigning the resultant DataFrame to a variable called `pca_with_metadata`."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":840},"executionInfo":{"elapsed":6,"status":"ok","timestamp":1704406056826,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"KzhDrDvtZXlI","outputId":"73628f45-8d42-401d-e563-0e8713fb38cc"},"outputs":[],"source":["### write your code below ###\n"]},{"cell_type":"markdown","metadata":{"id":"0DgEfjTIe3Dw"},"source":["Once you're ready, create the PCA scatterplot just as you did before: this time, use the `hue` parameter to color the PCA plot by `cell_ontology_class`."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":444},"executionInfo":{"elapsed":550,"status":"ok","timestamp":1704406057372,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"uGFY9LecjPOv","outputId":"e4a4239b-eea6-4562-eaf0-b8f2df1cb111"},"outputs":[],"source":["### write your code below ###\n"]},{"cell_type":"markdown","metadata":{"id":"8SkzDM0jjBo4"},"source":["### Gene expression\n","\n","Let's take this one step further by coloring our plot with gene expression values from `facs`.\n","\n","The authors of the Tabula Muris Senis [paper](https://www.nature.com/articles/s41586-020-2496-1) identified the top 20 genes upregulated and downregulated with age in three different tissues, including the bladder (Fig. 2F). Below, we've transcribed the list of genes from the figure for your convenience."]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":5,"status":"ok","timestamp":1704406057372,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"A1-ICijFL_PI"},"outputs":[],"source":["age_up = ['Sprr1a', 'Ly6d', 'Lgals3', 'Wfdc2', 'Gsta4', 'Gsto1', 'Krt18', 'Sfn',\n","          'Krt8', 'Areg', 'Aqp3', 'Igfbp2', 'Krt15', 'Gstm1', 'Krt19', 'Fxyd3',\n","          'Akr1b8', 'Foxq1', 'Cldn4', 'Krt7']\n","\n","age_down = ['Pcolce', 'Ppic', 'Id3', 'Col6a2', 'Gadd45g', 'Tuba1a', 'Fos', 'Dcn',\n","            'Col5a2', 'Htra1', 'Car3', 'Rcn3', 'Mgp', 'Lgals1', 'Col6a1',\n","            'Col1a1', 'Serpinh1', 'Sparc', 'Col1a2', 'Col3a1']"]},{"cell_type":"markdown","metadata":{"id":"vJT8vpcLCAJY"},"source":["Although our dataset is very high-quality, it's still useful to perform little sanity checks to make sure that everything is as it should be: in this case, we should make sure that all of the genes listed in the figure are actually in `facs`.\n","\n","Let's take this opportunity to review two core concepts: writing custom functions and using `for` loops."]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":4,"status":"ok","timestamp":1704406057372,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"vKdTbhVeMryB"},"outputs":[],"source":["# write a custom function called gene_check() that fulfills the following requirements:\n","# 1) takes two inputs: a list of genes (gene_list), and a DataFrame (df)\n","# 2) uses a for loop to check if each gene in age_up exists in df\n","# 3) if the gene is not in df, prints a warning message of your choice that\n","# includes the gene name\n","\n","### write your code below ###\n"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":225,"status":"ok","timestamp":1704406057593,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"vquiXtcOaou2","outputId":"75c37ac2-aeff-49d3-f584-be7643d40393"},"outputs":[],"source":["# use this cell to test your function\n","# none of the genes in control_genes should be found in facs\n","# all of the genes in age_up and age_down should be found in facs\n","\n","control_genes = pd.Series(['BRCA1', 'APOE', 'LCT'])\n","control_genes.index = control_genes\n","\n","gene_check(control_genes, facs)"]},{"cell_type":"markdown","metadata":{"id":"V-ufo3PGCJBk"},"source":["Now that we've verified that all of our aging-upregulated and downregulated genes are present in `facs`, we can color our PCA plot by individual genes and their expression values.\n","\n","Just as you did before, select the first two PCs from `pca_df` and join them with `facs` using `.merge()`, assigning the resultant DataFrame to `pca_with_expression`."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":562},"executionInfo":{"elapsed":182,"status":"ok","timestamp":1704406057773,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"9SAg2sy4lyph","outputId":"3d5c1ee0-9cd2-4dd3-c35f-8df3a2709d9d"},"outputs":[],"source":["### write your code below ###\n"]},{"cell_type":"markdown","metadata":{"id":"WmdiAETjquKc"},"source":["Once you're done, you can color your PCA plot with expression values from `age_up` and `age_down`, just like you did with the metadata.\n","\n","This excerpt from the article suggests some specific genes of interest:\n","\n","> \"*... using differential gene expression analysis to assess overall changes in tissues with age, we found that stromal-associated genes (Col1a1, Col1a2, Col3a1 and Dcn) are downregulated while epithelial-associated genes (Krt15, Krt18 and Sfn) are upregulated, supporting the compositional observations.*\"\n","\n","Notes:\n","* Set the `palette` parameter with a continuous palette of your choice: you can view different options [here](https://www.practicalpythonfordatascience.com/ap_seaborn_palette).\n","* For most palettes, it's advisable to also use the optional `legend = None` and `linewidth = 0` parameters to reduce visual noise.\n","* Expression values will be plotted with a wide color spectrum: we recommend sorting the DataFrame by the gene's expression values.\n","  * This ensures that the \"brightest\" points are plotted last, making it easier to see patterns in the colored plot.\n","\n","If you want to change additional aspects of your plot, you can view the documentation for `sns.scatterplot()` [here](https://seaborn.pydata.org/generated/seaborn.scatterplot.html).\n"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":466},"executionInfo":{"elapsed":573,"status":"ok","timestamp":1704406058343,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"Wrpswl2aqblh","outputId":"e2105d90-91b3-4111-e620-fd4b2cc3be33"},"outputs":[],"source":["### write your code below ###\n"]},{"cell_type":"markdown","metadata":{"id":"laRXV9sjU3pR"},"source":["## Final steps\n"]},{"cell_type":"markdown","metadata":{"id":"fONk44SRynMu"},"source":["### Saving plots\n","\n","`pyplot` offers a convenient function for saving plots. `plt.savefig()` requires a filename (`fname`) and takes an optional `dpi` parameter. We recommend using `dpi = 300` for publication-ready plots.\n","\n","To save your plot, simply add `plt.savefig()` to the end of your plotting cell, then run the cell."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":448},"executionInfo":{"elapsed":591,"status":"ok","timestamp":1704406058931,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"Q4qAZLqfzACE","outputId":"a3c4fefc-2dcc-4828-b464-6e1ccb57bfef"},"outputs":[],"source":["### place the code for your favorite plot below ###\n","\n","### place the code for your favorite plot above ###\n","\n","plt.savefig('my_favorite_plot.png', dpi = 300)"]},{"cell_type":"markdown","metadata":{"id":"WaW1dg8mzVLA"},"source":["You can now download this plot from the Files menu. (If it hasn't appeared, try refreshing the Files menu.)"]},{"cell_type":"markdown","metadata":{"id":"7Y9j1ur-1q3L"},"source":["### The ultimate test: comparing to `scanpy`\n","\n","There are many, many packages available for single-cell data analysis in Python: we just installed `scanpy`, a very popular package that was used in the full Tabula Muris Senis dataset analysis.\n","\n","Packages like `scanpy` are well-loved because they bundle together many single-cell operations and analyses into individual functions. Below, we've written the `scanpy` commands that replicate each step of what we did above, down to the final PCA plot."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":13527,"status":"ok","timestamp":1704406072453,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"_hPQzwaWA2o-","outputId":"2def0edf-ed92-4531-f6f4-08bf467ab4fd"},"outputs":[],"source":["# we need to use pip to install scanpy first\n","# this should take < 30 seconds\n","!pip install scanpy"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":448},"executionInfo":{"elapsed":20445,"status":"ok","timestamp":1704406092895,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"YnPuE29EJrCq","outputId":"c78beb91-8e3a-4fcd-ec11-6b518b3a7377"},"outputs":[],"source":["import scanpy as sc\n","import anndata\n","\n","# creates an object called adata that \"holds\" both the count matrix and metadata\n","# it's basically a container for various DataFrames in the dataset\n","adata = anndata.AnnData(X = facs)\n","adata.obs = metadata_facs\n","\n","# calculates highly variable genes and updates adata\n","sc.pp.highly_variable_genes(adata)\n","\n","# performs PCA and updates adata\n","sc.tl.pca(adata)\n","\n","# creates a PCA plot and colors it by one of our genes of interest\n","# it automatically sorts gene expression and places brightest points last\n","sc.pl.pca(adata, color='Krt15')"]},{"cell_type":"markdown","metadata":{"id":"oLHDiF1y3ywO"},"source":["Hopefully this looks similar to your PCA plot when colored by the same gene. If so, **congratulations**: you've matched the power of one of the most popular Python single-cell analysis packages!\n","\n","> *Note*: Colorbars aren't available in `seaborn`, only `pyplot` üò¢ We wanted to emphasize `seaborn` today, but you can overcome this aesthetic difference simply using `plt.scatter()` and `plt.colorbar()` in place of `sns.scatterplot()`. Don't let that detract from your victory over Python!"]},{"cell_type":"markdown","metadata":{"id":"G1T_eKG24ki2"},"source":["# Moving on with Python\n","This marks the end of PyCamp (üò¢), but hopefully not the end of your Python journey (üêç).\n","\n","The best way for you to get better with Python is to find a way to utilize it in your existing data analysis workflow. Although it might be a little bit of a slow start as you figure out how to build your analyses, practice makes perfect!\n","\n","If you'd like tutorials that go beyond the `numpy`/`pandas`/`pyplot`/`seaborn` content covered in PyCamp, consider working through the materials available on [DataQuest](dataquest.io) or [Calm Code](https://calmcode.io/). Both websites provide walkthroughs and exercises for specific data tasks, and you may find them helpful for reinforcing and filling in any gaps that exist post-PyCamp.\n","\n","Alternatively, if you're more of a book person, we recommend the following interactive books that are *also* published as physical books. You can order these via your favorite local/online bookshop.\n","* **Python for Data Analysis: Data Wrangling with Pandas, NumPy, and Python**: [online version](https://github.com/wesm/pydata-book); [publisher website](https://www.oreilly.com/library/view/python-for-data/9781491957653/)\n","* **Python Data Science Handbook**: [online version](https://github.com/jakevdp/PythonDataScienceHandbook), [publisher website](https://www.oreilly.com/library/view/python-data-science/9781491912126/)\n","* **Mastering Python for Bioinformatics: How to Write Flexible, Documented, Tested Python Code for Research Computing**: [online version](https://github.com/kyclark/biofx_python), [publisher website](https://www.oreilly.com/library/view/mastering-python-for/9781098100872/)\n","\n","All of the packages below have extensive tutorials and documentation, so you'll be able to transition fairly seamlessly from the PyCamp style tutorial of \"watch and repeat\" to working on your own code!\n","\n","* [SciPy](https://scipy.org/): A package that expands upon `numpy`, providing fundamental algorithms & functions for mathematical and statistical operations.\n","  * [scikit-image](https://scikit-image.org/): An affiliated package that provides image processing algorithms & functions.\n","  * [scikit-learn](https://scikit-learn.org/stable/): An affiliated package that provides machine learning algorithms and functions, one of which we used today for PCA!\n","* [Biopython](https://biopython.org/): A package for parsing and working with protein & sequence data, ranging from sequence parsing to alignment and querying of public databases.\n","* [statsmodels](https://www.statsmodels.org/stable/index.html): A module for statistical modeling and tests, expanding upon data structures from `numpy` and `pandas`.\n","* [keras](https://keras.io/): A relatively user-friendly package for deep learning."]},{"cell_type":"markdown","metadata":{"id":"JYDXCNjadIyb"},"source":["# Challenge zone\n","\n","The prompts below will require some more advance `pyplot`, `seaborn` and `pandas` techniques that we haven't covered explicitly in PyCamp. If you've finished early and would like to push yourself a bit, try them out: **note that we will not cover these solutions in session**."]},{"cell_type":"markdown","metadata":{"id":"nsP5DUXZr0bE"},"source":["## Multi-plot plots\n","\n","Let's say that we want to examine many different genes and their expressions within the figure. We can create multi-plot figures using `pyplot` and `seaborn` together.\n","\n","Multi-plot figures in `pyplot` are created with two types of objects.\n","* **Figure**: A Figure object is the \"base\" of your figure, akin to a blank canvas.\n","* **Subplot**: A Subplot is an object that only exists relative to a Figure: it specifies where your plotted data should go on the Figure.\n","\n","Below is an example image that shows what a Figure object with one subplot looks like.\n","\n","![fig_map](https://github.com/ccbskillssem/pythonbootcamp/raw/main/day_5/fig_map.png)\n","\n","The simplest way to get started is to use the `plt.subplots()` function, which creates a grid of `(nrows, ncols)` subplots. This function takes several parameters:\n","* `nrows`: The number of rows in your grid of subplots.\n","* `ncols`: The number of columns in your grid of subplots.\n","* `figsize`: As described for `plt.figure()` earlier, a tuple of `(width, height)`.\n","* `layout`: Configures orientation of subplots. We recommend setting `layout = 'tight'` to minimize whitespace between subplots.\n","\n","`plt.subplots()` will return a Figure and an array of Subplot objects, which are typically assigned to variables called `fig` and `axes`."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":462},"executionInfo":{"elapsed":2151,"status":"ok","timestamp":1704406095042,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"65pDrq_mvmaT","outputId":"1c18b667-68e5-426a-ed48-9b9b18900430"},"outputs":[],"source":["# example: a Figure with four Subplots\n","fig, axes = plt.subplots(nrows = 2, ncols = 2, figsize = (4, 4), layout = 'tight')\n","\n","print('Figure:', fig)\n","print('Subplots:', axes)"]},{"cell_type":"markdown","metadata":{"id":"UbU4ymt8wy4J"},"source":["The code below creates a Figure with a 2 x 2 grid of Subplots. It takes advantage of two new functions/methods:\n","* `.flatten()`: A method for arrays, which converts `axes` from a 2 x 2 array to a 1 x 4 array that we can use in a `for` loop.\n","* `zip()`: A built-in function that \"zips\" together iterables (list, array, etc) of equivalent length. The first gene is matched with the first subplot, and so on and so forth."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":807},"executionInfo":{"elapsed":3986,"status":"ok","timestamp":1704406099025,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"6-t7TZfUsTMt","outputId":"0ab0a37f-cd71-48d4-812c-5540d96be90b"},"outputs":[],"source":["# example: what if we wanted to create a Figure of 2x2 subplots\n","# each containing a PCA plot colored by a different gene?\n","# this will take a moment to generate\n","\n","plot_genes = ['Fxyd3', 'Gadd45g', 'Krt7', 'Col3a1']\n","\n","fig, axes = plt.subplots(nrows = 2, ncols = 2, figsize = (8, 8), layout = 'tight')\n","axes = axes.flatten()\n","\n","for gene, ax in zip(plot_genes, axes.flatten()):\n","  sns.scatterplot(pca_with_expression.sort_values(gene), x = 'PC1', y = 'PC2',\n","                  hue = gene, ax = ax,\n","                  palette = 'viridis', linewidth = 0, s = 11, legend = None)\n","  ax.set_title(gene)"]},{"cell_type":"markdown","metadata":{"id":"riFWFTZM0NHM"},"source":["Using the above code as a skeleton, create a Figure with a subplot for each gene in `age_up` (or `age_down`). You will need to adjust the grid dimensions and Figure size.\n","\n","We recommend making these changes as well:\n","* Use `.sort_values()`, as described earlier, to sort the DataFrame by the gene's expression values.\n","* Add a title to each subplot: at minimum, it should contain the gene name.\n","* Disable the legend in `sns.scatterplot()`: you can see how busy the plot is already!"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":1000},"executionInfo":{"elapsed":26605,"status":"ok","timestamp":1704406125626,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"0EhteVlB1_fb","outputId":"11124266-2624-47a8-ffb3-28807ba3bc33"},"outputs":[],"source":["### write your code below ###\n","### plots will take a while to generate! ###\n"]},{"cell_type":"markdown","metadata":{"id":"WIpbzs-cdiML"},"source":["## Changes in relative cell proportions with age (Fig. 2E)\n","\n","The Tabula Muris Senis dataset contains data derived from two different single-cell techniques:\n","\n","| Technique   | Quantity of cells | Sensitivity |\n","| ----------- | ----------------- | ----------- |\n","| Microfluidic droplet-based sequencing      | +++       | +       |\n","| FACS-sorted plate-based sequencing   | +        | +++       |\n","\n","Each technique has its own pros and cons: in this mini-project, we worked on the FACS data because its count matrix is smaller and less likely to overlab Colab's memory (and destroy your work in progress!)\n","\n","Let's try to replicate the figure and analysis of Figure 2E (\"Bladder cell\"; \"Bladder urothelial cell\"), which uses the larger droplet sequencing dataset. The figure and relevant section from the paper are reprinted below for your convenience.\n","\n","> *The cellular composition of each tissue type tends to vary with age. We investigated changes in the cellular composition of tissues for which data from at least three time points was available ...<br><br>The bladder shows pronounced changes in cell-type composition with age (Fig. 2e): whereas the mesenchymal compartment of this tissue decreases by a factor of three over the lifetime of the mouse (Fig. 2e, left), the urothelial compartment increases by a similar amount (Fig. 2e, right). The observation that the proportion of bladder urothelial cells increases with age is concordant with known age-related urothelial changes [25].*\n","\n","\n","<img src=\"https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41586-020-2496-1/MediaObjects/41586_2020_2496_Fig2_HTML.png?as=webp\">"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":304,"status":"ok","timestamp":1704406125928,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"v0_6X86yzQei","outputId":"f84411c1-74cc-4b7a-9b28-51429b477ccb"},"outputs":[],"source":["# run this cell to download the droplet metadata: this should only take a few seconds\n","%%bash\n","wget https://cells.ucsc.edu/tabula-muris-senis/droplet/bladder/meta.tsv -O bladder-droplet.meta.tsv --no-check-certificate --no-verbose"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":400},"executionInfo":{"elapsed":8,"status":"ok","timestamp":1704406125928,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"aGmW4ldGdon5","outputId":"f69eba74-24d8-4f31-88ab-7aaefc080e13"},"outputs":[],"source":["# use the appropriate pandas function to load in bladder-droplet.meta.tsv\n","# and assign it to a variable called metadata_droplet\n"]},{"cell_type":"markdown","metadata":{"id":"kxQwHnuAoI_U"},"source":["Figure 2E shows that the relative proportion of bladder cells decreases with age, whereas the relative proportion of bladder urothelial cells increases with age. Both trends are statistically significant.\n","\n","In order to create a plot-friendly DataFrame, you will need to do the following:\n","* Get group-based counts of cells by `age` and `cell_ontology_class` as a DataFrame. (Make sure it's a DataFrame and not a Series!)\n","* Create a new column with relative proportions of cell counts for each age group. **Make sure to check that the relative proportions for each sample sum up to 1.**\n","  * *Hint*: The `.transform()` method is very powerful for \"filling in\" values to each row that were originally calculated with grouped rows. Relevant StackOverflow answers for calculating relative proportions are likely to suggest that you use `.transform()`.\n","* Replace the strings in ages with `int` or `float` equivalents (ex. `3` instead of `3m`). This will make your data easier to plot.\n","* When you're confident that everything looks the way it should, reset the DataFrame index. This returns the data to a non-hierarchical format that `seaborn` and `pyplot` can parse."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":1000},"executionInfo":{"elapsed":7,"status":"ok","timestamp":1704406125928,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"5v02EuqqdqV9","outputId":"6b79c8ea-1545-45bc-c6bf-159ad55cae7c"},"outputs":[],"source":["### write your code below ###\n"]},{"cell_type":"markdown","metadata":{"id":"axTD2kYAVcsR"},"source":["Finally, let's create the plot(s).\n","\n","You'll need to use `sns.regplot()`, which will draw both a scatterplot and linear regression line. You can view the documentation for `sns.regplot()` [here](https://seaborn.pydata.org/generated/seaborn.regplot.html).\n","\n","Notes:\n","* You can draw separate plots for `'Bladder cell'` and `'Bladder urothelial cell'` or use `plt.subplots()` to create a single Figure with two plots.\n","* You can alter the regression line style to a black dashed line using `line_kws = {'linestyle': 'dashed', 'color': 'black'}`.\n","* For this exercise, ignore the different scatter point colors."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":428},"executionInfo":{"elapsed":7,"status":"ok","timestamp":1704406125928,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"uLkPF3H7WRl8","outputId":"4e366b52-0887-445d-92d7-2127128567f1"},"outputs":[],"source":["### write your code for the Bladder cell plot below ###\n"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":428},"executionInfo":{"elapsed":182,"status":"ok","timestamp":1704406126105,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"XySKO3d-b8nW","outputId":"443775bb-78ad-4dad-ee3c-e0ea283acfe9"},"outputs":[],"source":["### write your code for the Bladder urothelial cell plot below ###\n"]},{"cell_type":"markdown","metadata":{"id":"Gey29fZgbucD"},"source":["As a final touch, we can repeat the statistical analysis described below.\n","\n","> *The relative abundances of bladder cells ... and bladder urothelial cells ... change significantly with age. P < 0.05 and r2 > 0.7 for a hypothesis test with the null hypothesis that the slope is zero, using two-sided Wald test with t-distribution of the test statistic.*\n","\n","We can perform this test using `scipy.stats.linregress()`. The documentation for this function is [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.linregress.html). You will need to perform this test once for `Bladder cells`, then again for `Bladder urothelial cells`."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":4,"status":"ok","timestamp":1704406126106,"user":{"displayName":"Stacy Li","userId":"17654991658950232542"},"user_tz":480},"id":"c5ta3OLcbgDR","outputId":"030c9453-25eb-491a-c0bf-a7353172bd59"},"outputs":[],"source":["from scipy import stats\n","\n","### write your code below ###\n"]}],"metadata":{"colab":{"authorship_tag":"ABX9TyM6cIYZrNW7XxQVpIn4Z5OM","provenance":[],"toc_visible":true},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"name":"python"}},"nbformat":4,"nbformat_minor":0}
