{"cells":[{"cell_type":"markdown","metadata":{"id":"KBpjUiS8duMv"},"source":["# Data exploration with DataFrames, cont.\n","\n","<font color = '#ed865c' size = 4>**Make sure to press the play button to run the cell below: this will re-load the datasets and functions that we worked on during the last section.**</font>"]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":18328,"status":"ok","timestamp":1704232955403,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"SD9-mk1MUw_h","outputId":"f03e1abd-ad53-4f4b-d7c0-94fa04293d44"},"outputs":[],"source":["#@title\n","# make sure to run this cell to import the external files we need for today\n","# and load in the appropriate packages\n","!git clone https://github.com/ccbskillssem/pythonbootcamp.git\n","\n","import numpy as np\n","import pandas as pd\n","import matplotlib.pyplot as plt\n","\n","alleles = pd.read_table('/content/pythonbootcamp/day_4/alleles.tsv')\n","clinvar = pd.read_table('/content/pythonbootcamp/day_4/clinvar.tsv')"]},{"cell_type":"markdown","metadata":{"collapsed":true,"id":"omuk9pm19NbP"},"source":["## Changing labels\n","\n","<font color = '#ed865c' size = 4>**Make sure to run the cell above before you begin: this will re-load the datasets and functions that we worked on during the last section.**</font>"]},{"cell_type":"markdown","metadata":{"id":"77qtOEDXXTVt"},"source":["### Row labels\n","\n","Currently, the row labels for `alleles` are numeric, which makes it easy to index rows just as we did with arrays. However, it's also possible to use row *names* as labels, just as we do with columns.\n","\n","To do this, we can select an existing column of the DataFrame and turn it into the row Index. However, it's **extremely** important that we make sure the candidate column contains only unique values: we wouldn't want to try and index a row name, only to retrieve multiple rows with the same name!\n","\n","Series have a useful attribute called `.is_unique`, which is `True` if all values in the Series are unique. Given that columns of DataFrames are stored as Series, we can easily check our candidate columns."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":219,"status":"ok","timestamp":1704232955619,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"kbecHau7V-kc","outputId":"d3950338-c7c7-4150-893f-e142a6aa9478"},"outputs":[],"source":["# try it out:\n","# print the .is_unique attribute of each column\n","# hint: remember for loops? try using a loop with the column labels\n"]},{"cell_type":"markdown","metadata":{"id":"4eYKTPWKYoNJ"},"source":["Great – looks like there's a candidate column for our row labels. We can now change `allele`'s row labels using the `.set_index()` method."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":237},"executionInfo":{"elapsed":363,"status":"ok","timestamp":1704232955981,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"q5Y3JAViV-nr","outputId":"06dec454-bc38-426a-d82e-14f212f6df77"},"outputs":[],"source":["# in this case, we're opting to update the variable\n","\n","alleles = alleles.set_index('ID')\n","alleles.head()"]},{"cell_type":"markdown","metadata":{"id":"7gBXnI5ZLgTp"},"source":["Excellent! As you can see, each row's `ID` value is now the row label. We can access each row using `.loc[]` and the corresponding `ID` label."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"collapsed":true,"executionInfo":{"elapsed":182,"status":"ok","timestamp":1704232956161,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"VeDu6GfD9NbQ","outputId":"75c0d9b4-b33e-4cc2-8ab3-1d41c8140ca2"},"outputs":[],"source":["alleles.loc['rs573303859']"]},{"cell_type":"markdown","metadata":{"id":"KM9jsVS8azTp"},"source":["Row names can be convenient, but also cumbersome: for example, now that we've assigned `ID` as our row labels, all of the row labels are strings: thus, we can no longer slice multiple rows with `.loc[]`.\n","\n","> If you really want to slice multiple rows *and* retain string-type row labels, you will need to use `.iloc[]` instead of `.loc[]`. `.iloc[]` is used to access columns/rows/slices/cells in a `numpy` array-like manner. You can read more about it [here](https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#selection-by-position)."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"collapsed":true,"executionInfo":{"elapsed":147,"status":"ok","timestamp":1704232956486,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"WdSLWVFt9NbR","outputId":"57d4c572-b90c-45b8-ef30-481770060004"},"outputs":[],"source":["# try it out:\n","# access the row labeled rs563299082\n","# and its chromosome, position, reference, and alternative values\n"]},{"cell_type":"markdown","metadata":{"id":"jVWbmjVYPk9I"},"source":["If you end up wanting to revert back to the standard 0-index numeric row labels, you can use the `.reset_index()` method to convert the row names back to a regular column."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":206},"executionInfo":{"elapsed":200,"status":"ok","timestamp":1704232956842,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"z3oNy_G8Ps7s","outputId":"2b425ae1-7bdd-4fd0-cd8a-c1626099955a"},"outputs":[],"source":["# let's return alleles to its original state\n","\n","alleles = alleles.reset_index()\n","alleles.head()"]},{"cell_type":"markdown","metadata":{"id":"x5oOlBezKhvg"},"source":["### Column labels\n","In the case that we want to replace column labels, we can use the `.rename()` method, which is the equivalent of `.replace()` for column labels. Just as we showed you before, the easiest way to do this is to provide a dictionary of target and substitute labels.\n","\n","For example, let's consider `reference` and `alternative`, which are commonly shortened to `REF` and `ALT` in genomic file formats. Changing the column labels while maintaining the column content is a single-line operation:"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":206},"executionInfo":{"elapsed":239,"status":"ok","timestamp":1704232957427,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"qfkqn7lGK2fK","outputId":"ef681d8a-9882-4892-8fde-d4a05f569f74"},"outputs":[],"source":["# rename reference and alternative to REF and ALT for all rows\n","\n","alleles.rename({'reference': 'REF',\n","                 'alternative': 'ALT'}, axis = 'columns').head()"]},{"cell_type":"markdown","metadata":{"id":"d-rGpELa9NbY"},"source":["## Sorting by column(s)\n","There's no way to sugar-coat it: there was no intuitive way for us to sort arrays by specific columns in `numpy`.\n","\n","With `pandas`, the `.sort_values()` method allows us to sort a DataFrame/Series with the following inputs:\n","* A single column name, or a list of column names\n","* The `ascending` parameter, which takes a single Boolean value or a list of Boolean values (corresponding to the list of column names).\n","  * `True` will sort by lowest to highest (ascending).\n","  * `False` will sort by highest to lowest (descending)."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":206},"executionInfo":{"elapsed":191,"status":"ok","timestamp":1704232958340,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"78hX4GL39NbY","outputId":"82c5174d-e6fe-4a7b-b945-bbd8a0ce2467"},"outputs":[],"source":["# let's sort AFR_AF by highest to lowest frequency\n","\n","alleles.sort_values(\"AFR_AF\", ascending=False).head()"]},{"cell_type":"markdown","metadata":{"id":"CsWhTAk9LbXh"},"source":["⏸ **Exercise 1** Try sorting `alleles` by *all three* of the following columns:\n","* `AFR_AF`, from highest to lowest.\n","* `AMR_AF`, from lowest to highest.\n","* `EAS_AF`, from highest to lowest."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":362},"executionInfo":{"elapsed":479,"status":"ok","timestamp":1704232958980,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"rGrU-eqmLlSb","outputId":"e0a83b6a-de42-4421-fc5f-525512bde525"},"outputs":[],"source":["### write your code below ###\n"]},{"cell_type":"markdown","metadata":{"id":"EQwPqwWJe3Xi"},"source":["## Adding columns\n","\n","We can easily create new columns by simply assigning a new column name and contents to a DataFrame."]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":212,"status":"ok","timestamp":1704232959953,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"eXA4Mhdswe7T"},"outputs":[],"source":["# let's sample some rows with .sample()\n","# using 2024 as our seed value\n","\n","some_alleles = alleles.sample(n = 5, random_state = 2024)"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":206},"executionInfo":{"elapsed":306,"status":"ok","timestamp":1704232960712,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"2Gu5Hr7UfPoV","outputId":"b3a50a0d-7f8a-42ae-f15b-d30e5d63cc7a"},"outputs":[],"source":["some_alleles['sample_col'] = 0 # a new column full of zeros\n","some_alleles"]},{"cell_type":"markdown","metadata":{"id":"pdP5qBs_fxMs"},"source":["---\n","⏸ **Exercise 2 (No coding)** What happens when you change the seed value of a random sampler (like `.sample()`)? What happens if you don't have a seed value specified?\n","\n","Hint: One of the morning exercises might help you answer this question.\n","\n","---\n","\n","Next, we can also assign new columns by explicitly providing a column label and a \"definition\" for the column."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":206},"executionInfo":{"elapsed":5,"status":"ok","timestamp":1704232961401,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"6SOaGBXwf1BI","outputId":"dbf79bb4-73c2-49a3-9c9e-782b0b3f4a64"},"outputs":[],"source":["# let's reuse some_alleles\n","# creating a new column that contains the *reference* allele frequency\n","# this takes advantage of numpy broadcasting\n","\n","some_alleles['reference_AF'] = 1 - some_alleles['AF']\n","some_alleles"]},{"cell_type":"markdown","metadata":{"id":"ebxxQU9P46Ym"},"source":["Lastly, the `.map()` method allows us to **mutate** new columns based on values of existing columns. This is useful for adding categorical or Boolean columns without overwriting existing columns."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":206},"executionInfo":{"elapsed":406,"status":"ok","timestamp":1704232962252,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"-WH3ZwgN43Dt","outputId":"3a3de88a-2b79-4ef6-8f79-3f468233d907"},"outputs":[],"source":["# let's mutate a new column called filter_pass\n","# instead of replacing the values of filter\n","\n","alleles['filter_pass'] = alleles['filter'].map({'PASS': True,\n","                                      'LowQ': False})\n","alleles.head()"]},{"cell_type":"markdown","metadata":{"id":"uJl_v7MhWu9J"},"source":["## Dropping columns\n","\n","We can use the `.drop()` method to drop columns as easily as we created them. This can be handy if you end up with too many columns during your exploration, or if you simply don't need some of the columns that came with your dataset."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":423},"executionInfo":{"elapsed":226,"status":"ok","timestamp":1704232962894,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"U5vqgpqxZaoV","outputId":"faafc2d6-be63-4799-d29e-44c38af63fa5"},"outputs":[],"source":["# getting rid of filter_pass\n","# removing it from all rows in the dataframe\n","\n","alleles = alleles.drop('filter_pass', axis = 'columns')\n","alleles"]},{"cell_type":"markdown","metadata":{"id":"jsjrmgA_Y-LC"},"source":["## Grouping\n","\n","We've now reached the operations that are *only* possible in `pandas`. **Grouped operations** partition the dataset into subsets based on the value of a given column or columns prior to operation.\n","\n","There are two major steps in performing grouped operations:\n","1. **Splitting**: Splitting data into groups.\n","2. **Applying**: Applying a function across separate groups. (This is where vectorization is helpful!)\n","\n","We can accomplish the first step using the `.group_by()` method, which takes one or more column labels."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":203,"status":"ok","timestamp":1704232964137,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"yCLD-oZoY-LD","outputId":"44947931-e8e6-4783-ac89-9841c62210db"},"outputs":[],"source":["# this method doesn't show a visible output\n","\n","alleles.groupby('filter')"]},{"cell_type":"markdown","metadata":{"id":"-aQMpDbrY-LD"},"source":["In order to see the effect of splitting with `.groupby()`, we need to chain `.groupby()` with a function that we want applied to our groups.\n","\n","For example, let's say that we want to know the average quality score of alleles that pass (or don't pass) filter."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":179,"status":"ok","timestamp":1704232964502,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"YjgOxlEOY-LD","outputId":"142bdd4a-184d-4c4e-f1e4-6772c34bbb23"},"outputs":[],"source":["alleles.groupby('filter')['quality'].mean()"]},{"cell_type":"markdown","metadata":{"id":"0OJluYxEIIAe"},"source":["## Merging (review)\n","\n","This morning we discussed merging two datasets: `alleles` and `clinvar`. Merging is a very useful operation that we'll be using in tomorrow's final mini-project.\n","\n","Once again, merge `alleles` and `clinvar` so that we have a single DataFrame called `merged`, which contains only the alleles present in both of the datsets."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":513},"executionInfo":{"elapsed":1216,"status":"ok","timestamp":1704232971778,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"YQlKgG9yKN7V","outputId":"6ec807e2-c83e-4f80-e472-a856568a25b5"},"outputs":[],"source":["# try it out:\n","# merge alleles and clinvar as you did earlier today\n"]},{"cell_type":"markdown","metadata":{"id":"KSkXx3x2oiL6"},"source":["## Exporting\n","\n","This is our final lesson in `pandas`: hooray!\n","\n","We've learned a lot about how to parse and manipulate our DataFrames. At some point in the near future, you'll likely want to save the results of your operations.\n","\n","The `.to_csv()` method allows us to easily export our DataFrame to a `.csv` file. As the name implies, it will save a DataFrame to a `.csv` file, given a string (file name).\n","\n","Let's go ahead and cap off our lecture by exporting `merged` to a file called `merged.csv`."]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":2,"status":"ok","timestamp":1704232971778,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"jIKbtdJsod7A"},"outputs":[],"source":["merged.to_csv('merged.csv')"]},{"cell_type":"markdown","metadata":{"id":"bEKVQFWYonZ9"},"source":["If you open the `Files` menu (left panel, folder icon), you'll see that the `merged.csv` file is now available.\n","___\n","**CAUTION**: Files that you save while using Colab are not retained after you close the notebook, as they only exist in Colab's temporary **session storage**. If you generate files and wish to keep them, make sure to download your files (with the same three dots menu) before you exit Colab.\n","___\n","\n","That's it for our introduction to `pandas`! We hope you're starting to get a feel for how immensely powerful `pandas` can be. (Perhaps too powerful for a single bootcamp lecture...)\n","\n","* To learn more about DataFrame methods, click here for the DataFrame documentation.\n","* To learn more about Series methods, click here for the DataFrame documentation.\n","* To learn about other common `pandas` routines, refer to the official `pandas` cookbook [here](https://pandas.pydata.org/docs/user_guide/cookbook.html). (No pandas were harmed in the making of this cookbook.)\n"]},{"cell_type":"markdown","metadata":{"id":"QPFqPTYoOzbH"},"source":["# << `Exercises` >>\n","\n","⏸ **Exercise 3a**: Just like with arrays, we can use the `.isna()` method to detect missing values in a DataFrame. Use this to identify all missing values in `clinvar`."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":362},"collapsed":true,"executionInfo":{"elapsed":3,"status":"ok","timestamp":1704232977702,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"WYtQ7LYN9Nba","outputId":"4db45be7-25e4-48df-9105-9783a3716144"},"outputs":[],"source":["### write your code below ###\n"]},{"cell_type":"markdown","metadata":{"id":"UwdAoacnP8lw"},"source":["⏸ **Exercise 3b** Let's chain together some `pandas` methods! Which columns of `clinvar` are missing values?\n","\n","*Hint*: The `.any()` method returns a `True` if any of the values in the Series or array are `True`."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"collapsed":true,"executionInfo":{"elapsed":1770,"status":"ok","timestamp":1704232980519,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"h0wP-sgR9Nba","outputId":"44172bef-66b7-4e15-b463-c03137e91743"},"outputs":[],"source":["### write your code below ###\n"]},{"cell_type":"markdown","metadata":{},"source":["⏸ **Exercise 3c**: Find the total number of missing values in each column of `clinvar`.\n","\n","*Hint*: Recall that `False` Booleans are considered equivalent to `0`, and `True` values are considered equivalent to `1`."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["### write your code below ###\n"]},{"cell_type":"markdown","metadata":{"id":"d07XuqgLRIsH"},"source":["⏸ **Exercise 3d**: Which rows in `clinvar` are incomplete? Filter `clinvar` to show these rows.\n","\n","*Hint*: You can change the axis along which `.any()` operates."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":989,"status":"ok","timestamp":1704232981505,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"7eVGm7yQRPHJ","outputId":"7c524e9e-6cf5-4ed9-c35a-2aed9626d46d"},"outputs":[],"source":["### write your code below ###\n"]},{"cell_type":"markdown","metadata":{"id":"uT5lc3GHRZri"},"source":["⏸ **Exercise 4a** We're now going to move to data cleaning. \n","\n","First, select rows in ```clinvar``` that have complete data (no missing values): save this filtered DataFrame to a variable called `clinvar_complete`. What are the dimensions of `clinvar_complete`?"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"collapsed":true,"executionInfo":{"elapsed":767,"status":"ok","timestamp":1704232982270,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"rgM6y32q9Nbb","outputId":"46127d84-1b9e-41ac-cce8-6357061ede78"},"outputs":[],"source":["### write your code below ###\n"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"collapsed":true,"executionInfo":{"elapsed":682,"status":"ok","timestamp":1704232983906,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"wYeA_7Ef9Nbb","outputId":"117c3c22-5b93-474e-c4b7-dcd168b424c4"},"outputs":[],"source":["# use this cell to check your work\n","clinvar_complete.isna().any()"]},{"cell_type":"markdown","metadata":{},"source":["⏸ **Exercise 4b**: It seems like a significant number of `clinvar` rows contain missing values. Calculate the mean, median, and maximum number of missing values for `clinvar` rows.\n","\n","*Hint*: This exercise uses the same principle as Exercise 3c."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["### write your code below ###\n","\n"]},{"cell_type":"markdown","metadata":{"id":"aoUWEEYTTqJj"},"source":["⏸ **Exercise 4c**: The `.dropna()` method accomplishes what we did with Boolean filtering, but it allows for additional flexibility when filtering rows. Look up the documentation for `.dropna()`, then use it to select rows in `clinvar` that have no more than 2 NA values. Assign this DataFrame to `clinvar_partial`. "]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["### write your code below ###\n"]},{"cell_type":"markdown","metadata":{},"source":["⏸ **Exercise 4d**: How many rows remain compared to `clinvar_complete` and the original `clinvar` DataFrame?"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":1020,"status":"ok","timestamp":1704232987980,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"tGJVLvyMS5c7","outputId":"da1ad39d-47af-4b66-837e-42f21a3e7d2f"},"outputs":[],"source":["### write your code below ###\n"]},{"cell_type":"markdown","metadata":{"id":"pQaPlK2F9Nbd"},"source":["⏸ **Challenge**: Create a new DataFrame called `pathogenic` that contains only pathogenic variants.\n","\n","*Hint*: The `clinical_significance` column contains several different strings with the substring `'pathogenic'`. How can you take advantage of this to filter `clinvar`? If you're not sure, check the morning notebook.)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["### write your code below ###\n"]},{"cell_type":"markdown","metadata":{},"source":["Find the number of pathogenic variants for each of three genes: `BRCA1`, `PAH`, and `CFTR`."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["### write your code below ###\n"]},{"cell_type":"markdown","metadata":{"id":"KTjxVlR1pPPQ"},"source":["# [Optional] More about `pandas`\n","\n","Are you excited about `pandas`? Like, really excited? Here's some extra info on techniques that may be useful to you in the future. (We're not including this in the main lecture, as there's already a ton to go through!)\n","\n","We'll focus on the `merged` dataset, which we derived by merging together `alleles` and `clinvar`. All of the rows in `merged` represent alleles found in both datasets."]},{"cell_type":"markdown","metadata":{"id":"w1Fb2miFp0YD"},"source":["First, let's quickly inspect a couple of different things, just to get our bearings."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":150,"status":"ok","timestamp":1704232991050,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"Ka8iCiUjpc92","outputId":"20ca16f3-48b3-43e4-fbf2-fa6f71160b5b"},"outputs":[],"source":["# try it out:\n","# print the # of rows and cols in merged\n","# just to review how many alleles in alleles are also found in clinvar\n"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":412},"executionInfo":{"elapsed":129,"status":"ok","timestamp":1704232991707,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"1VG8dC_NGj28","outputId":"ac4dea9c-4081-4b51-d994-6f2fa8e816b9"},"outputs":[],"source":["# try it out:\n","# view the first 10 rows of merged\n"]},{"cell_type":"markdown","metadata":{"id":"waEI8pYdrIKC"},"source":["Let's quickly look at the column labels, just so we can have a list of the columns we might want to examine."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":129,"status":"ok","timestamp":1704232992930,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"w6GWJ-7UprIV","outputId":"6312525e-895d-4e7d-dd63-a0028bd7e76c"},"outputs":[],"source":["# try it out:\n","# examine the column labels: how many are there, and what are they?\n"]},{"cell_type":"markdown","metadata":{"id":"S7uIDOCbrkmC"},"source":["Great! We can see that we have a good number of columns that derive from `clinvar`. There are also some duplicate columns, which are marked with the `_x` suffix to indicate that they derive from `alleles`, or the `_y` suffix to indicate that they derive from `clinvar`. You can adjust these suffixes using parameters in `pd.merge()`.\n","\n","Let's examine some of the columns that may hold interesting categorical values:"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":2,"status":"ok","timestamp":1704232993366,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"7P71dJKarXog","outputId":"6de88043-928b-4fe7-ed5a-a83b7165ceda"},"outputs":[],"source":["# try it out:\n","# find the unique values in the clinical_significance column\n"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":2,"status":"ok","timestamp":1704232993530,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"6rC3ecQvr2La","outputId":"953431ef-62af-4c59-ef3a-f08e26e84b3a"},"outputs":[],"source":["# try it out:\n","# find the unique values in the molecular_consequence column\n"]},{"cell_type":"markdown","metadata":{"id":"2BBmn_c68b7M"},"source":["Given that not everyone will be versed in disease genetics or molecular biology, here's a quick crash course on what these mean.\n","\n","*clinical_significance* refers to the observed effect of the mutation on human health. This data is curated by clinicians and uploaded to the ClinVar database.\n","\n","*molecular_consequence* refers to the consequence of a specific mutation on the gene product (protein) of the mutated gene. Not all mutations yield the same molecular_consequence, and some consequences are more dire than others: for example, *nonsense mutations* can result in a gene product that may not perform its biological function."]},{"cell_type":"markdown","metadata":{"id":"-Fq2YWYnRKyu"},"source":["## Fun tricks: rolling window operations\n","\n","This is a fun feature of `pandas` that may be interesting to people who work with longitudinal-type data, such as time series data or genomic coordinate-related data. It is *not* a must-know feature of `pandas`.\n","\n","The `.rolling()` method for DataFrames takes a `window` parameter, which specifies the number of rows to use to create a *rolling (sliding) window* across the DataFrame. This can be used to methodically calculate summaries such as sliding window averages using a `for` loop. (Yes, this is a very specific trick!)"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":3394,"status":"ok","timestamp":1704232997417,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"V5vfFF-rQ9Tk"},"outputs":[],"source":["# calculating a rolling window average of allele frequencies on the 21st chromosome\n","# using the alleles dataset, sorted by genomic position on the 21st chromosome\n","\n","rolling_mean = []\n","for window in alleles[alleles['chromosome'] == 21].sort_values('position').rolling(window = 500):\n","  rolling_mean.append(window['AF'].mean()) # get the average allele freq in the rolling wondow"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":447},"executionInfo":{"elapsed":630,"status":"ok","timestamp":1704232998045,"user":{"displayName":"Sarah Ariana Johnson","userId":"18296486351826396058"},"user_tz":480},"id":"8HIn1KpISfWO","outputId":"128adca9-6e3f-451f-8265-396631becaea"},"outputs":[],"source":["# plotting the rolling mean of allele frequencies across chromosome 21\n","# plt.plot takes two arrays/lists: the first corresponds to x values, the second to y values\n","# below, we created \"dummy\" x values that match the length of rolling_mean\n","# (there are more robust ways to plot this rolling mean :) )\n","\n","plt.plot(np.arange(0, len(rolling_mean)), rolling_mean)"]},{"cell_type":"markdown","metadata":{"id":"sv1Z49qaU3Hf"},"source":["Want to learn more fun `pandas` tricks? It's time for you to move on to the full documentation [here](https://pandas.pydata.org/docs/user_guide/index.html#user-guide). Go wild!"]}],"metadata":{"colab":{"provenance":[{"file_id":"1VMdvmFXejAWMXB1PZKCpIYMuph0dmO2W","timestamp":1704053337538},{"file_id":"1RgqyHoS--aMDXovELYmUc-0BoKDswnez","timestamp":1655270967681}]},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.12.0"}},"nbformat":4,"nbformat_minor":0}
